### PR TITLE
Deploy declaration resource

### PR DESCRIPTION
**Pull request by zhaoqin-github**

### CODE REVIEW

File: app/waf/src/controllers/declaration.controller.ts [hunk 1]
```diff
@@ -333,4 +342,56 @@ export class DeclarationController extends BaseController {
       .declarations(applicationId)
       .delete({and: [{id: declarationId}, {tenantId: await this.tenantId}]});
   }
+
+  @post(
+    prefix +
+      '/applications/{applicationId}/declarations/{declarationId}/deploy',
```
Comment (zongzw) [inline app/waf/src/controllers/declaration.controller.ts]: As I recommended before, I don't think to prefix `/applications/{applicationId}` to the API is a good idea. 

Indeed, the declaration has a relation with the application... In fact, the declaration has a relationship with all elements of the TREE... 

The relation link has already been saved in declaration model, we/user can get it through the query string(`GET /declarations?applicationId=xxx` or `GET /applications/{applicationId}/delcarations`). 
That's not meaning we need to bind the relationship everywhere, even into the API call. 

Think that: when the application TREE's configuration is newer than a declaration generated before, to prefix the applicationId to that declaration's deployment would be confused, no matter succeed or fail. 

I think declaration is a standalone existence, which can exist or be operated on with no regarding the application which it is generated from...

File: app/waf/test/acceptance/declaration.controller.acceptance.ts [hunk 2]
```diff
@@ -385,6 +386,115 @@ describe('ApplicationController declaration test', () => {
...
+      prefix +
+      '/applications/{applicationId}/declarations/{declarationId}/deploy: deploy declaration',
+    async () => {
+      const adc = await givenAdcData(wafapp);
+      const application = await givenApplicationData(wafapp, {
+        adcId: adc.id,
+      });
+      const declaration = await givenDeclarationData(wafapp, {
+        applicationId: application.id,
+        tenantId: ExpectedData.tenantId,
```
Comment (zhang-shengping) [inline app/waf/test/acceptance/declaration.controller.acceptance.ts]: you may delete this `tenantId: ExpectedData.tenantId,`

File: app/waf/test/acceptance/declaration.controller.acceptance.ts [hunk 3]
```diff
@@ -385,6 +386,115 @@ describe('ApplicationController declaration test', () => {
...
+
+  it(
+    'post' +
+      prefix +
+      '/applications/{applicationId}/declarations/{declarationId}/deploy: deploy without default ADC',
+    async () => {
+      const application = await givenApplicationData(wafapp);
+      const declaration = await givenDeclarationData(wafapp, {
+        applicationId: application.id,
+        tenantId: ExpectedData.tenantId,
```
Comment (zhang-shengping) [inline app/waf/test/acceptance/declaration.controller.acceptance.ts]: the same as before

File: app/waf/src/controllers/declaration.controller.ts [hunk 4]
```diff
@@ -333,4 +342,56 @@ export class DeclarationController extends BaseController {
...
+
+    let application = await this.applicationRepository.findById(applicationId);
+
+    if (!application.adcId) {
+      throw new HttpErrors.UnprocessableEntity(
+        'No target ADC to perform deploy action',
+      );
+    }
+
+    let adc = await this.adcRepository.findById(application.adcId);
```
Comment (zhang-shengping) [inline app/waf/src/controllers/declaration.controller.ts]: do we need to select by tenantId here?

Comment (zongzw) [reply app/waf/src/controllers/declaration.controller.ts]: BTY,  can we replace find with findById at #L367?

**Comment by zhaoqin-github** [comment]
This change set is duplicated with PR #160. Let's commit all changes in #160 as a whole.

**Pull request closed without merging**